### DISCUSSION TITLE
Datagrid check if detailrow is displayed?

### DISCUSSION THREAD

**Discussion by svenbaus** [Megabit/Blazorise#4744] [Q&A]
Hello everybody,

I'm trying to use the datagrid in my application and want to have a Button that toggles a detailrow. The button image should change from ExpandMore to ExpandLess when the detailrow is displayed and back to ExpandMore when the detailrow is not displayed. Is there any possibility how to achieve this?

Some code:

```
<DataGrid @ref="dataGrid" TItem="ProcessFlow" Data="processFlows" Responsive Virtualize Editable="false" Filterable Resizable ResizeMode="TableResizeMode.Columns" SelectionMode="DataGridSelectionMode.Multiple" @bind-SelectedRows="selectedProcessFlows" DetailRowStartsVisible="false" ShowPager ShowPageSizes RowSelectable="RowSelectableHandler">
            <DataGridColumns>
                <DataGridMultiSelectColumn />
[...]
<DataGridColumn Caption="@_localizer["Details"]">
                    <DisplayTemplate>
                        <Button Color="Color.Secondary" Size="Size.Small" Clicked="() => OnDetailsButtonClicked(context)"><Icon Name="IconName.ExpandMore" /></Button>
                    </DisplayTemplate>
                </DataGridColumn>
            </DataGridColumns>
            <DetailRowTemplate>
                @{
                    //Select details only when nescearry
                    var dbContext = _dbContextFactory.CreateDbContext();
                    var processFlowWithDetails = dbContext.GetProcessFlowWithReferencesById(context.Id);
                    <Processdetails ProcessFlow="@processFlowWithDetails" />
                }
            </DetailRowTemplate>
        </DataGrid>

async Task OnDetailsButtonClicked(ProcessFlow? processFlow)
    {
        if ((dataGrid != null) && (processFlow != null))
        {
            await dataGrid.ToggleDetailRow(processFlow);
        }
    }
```

**Comment by David-Moreira** [comment]
If I remember correctly this event has information about the current state of the **DetailRow** when it's about to trigger: 
![image](https://user-images.githubusercontent.com/22283161/233095972-83a5087d-41de-4d40-8e68-1d6205ebbb03.png)

You should be able to achieve what you want with that, can you try it and let us know if it worked?

I will open an issue out of this to improve documentation for this use case, As it's a valid use case people will want, some Expand/Collapse Arrow that's synchronized with the DetailRow state.

**Reply by svenbaus** [reply]
Thanks for your answer. Well, I'm currently hanging, on how to use the DetailRowTrigger, since in my code example I use the method "ToggleDetailRow()".

**Reply by David-Moreira** [reply]
If I remember correctly it will be called everytime it evaluates whether to show or hide a `DetailRow` and it returns a boolean meaning it should show or not.

That's how originally it worked before v1.0. You gave it a condition and it would keep evaluating whether the condition would be true, to display the `DetailRow`. Nowadays it's a bit more imperative for performance reasons, but it should still work the same way it did.

We'll improve documentation on `DetailRow` feature usage as part of #4745

I believe you should be able to mesh both together, just make sure to return the appropriate bool everytime.
**PS:** @svenbaus see the note about `DetailRowTriggerType` & `Manual`.

---

Patch notes for more context.
https://blazorise.com/news/release-notes/100
![image](https://user-images.githubusercontent.com/22283161/233101581-81113186-384c-4910-907f-2f67257f92be.png)

**Reply by svenbaus** [reply]
I'm not quite shure, on how to get the Toggleable in some kind of if statement?! What I understand is, that there is a DetailRowTrigger, which exposes the Toggleable, but how do I get it here?

```
<DataGridColumn Caption="@_localizer["Details"]">
    <DisplayTemplate>
    @if(Toggleable)
    {
        <Button Color="Color.Secondary" Size="Size.Small" Clicked="() => OnDetailsButtonClicked(context)"><Icon Name="IconName.ExpandMore" /></Button>
    }
    else 
    {
        <Button Color="Color.Secondary" Size="Size.Small" Clicked="() => OnDetailsButtonClicked(context)"><Icon Name="IconName.ExpandLess" /></Button>
    }
    </DisplayTemplate>
</DataGridColumn>
```

**Reply by David-Moreira** [reply]
The `Toggleable` isn't meant to report the Expanded State. It does something else, as described in the patch notes.
I will take a look at your example and try to make it work, I'm positive it should work as I remember there being users implementing the same functionality before.

**Reply by David-Moreira** [reply]
Actually if you have time @svenbaus please see if anything helps you with implementing the feature in this issue : https://github.com/Megabit/Blazorise/issues/3226

As it might be a while before I get to try this. 
Please let me know if it helps and you figure it out.

**Reply by svenbaus** [reply]
Ok, I think I get, what you mean. The basic idea seems to be, that I save for myself, that the details row for this item has been triggered and therefore change the icon during rendering, correct? Its basically what I thought before asking, but I think this is very buggy when a new version of datagrid comes up.

**Reply by David-Moreira** [reply]
Hello @svenbaus 
I believe that's how that user did, yes. Not sure if there's a better way, because the state is basically hidden from you, so you need to track it somehow.

We've already talked briefly about the possibility of exposing something like this, that might help with these kind of things, but it's currently a private member
![image](https://user-images.githubusercontent.com/22283161/233427009-17a09443-9b44-4d53-a280-e1880e5296fe.png)


Can you clarify this statement? 
> but I think this is very buggy when a new version of datagrid comes up.

What's buggy? Is something not working correctly? Or do you just mean it's limited for what you're trying to do currently?
